### PR TITLE
fix(napi): Implement `napi_threadsafe_function` ref and unref

### PR DESCRIPTION
**Pull request by littledivy**
Closes https://github.com/denoland/deno/issues/17288

### CODE REVIEW

**Comment by littledivy** [comment]
AFAIU ref/unref do the same thing as release/aquire functions but the later also consumes/takes ownership of the tsfn handle.

File: cli/napi/threadsafe_functions.rs [hunk 1]
```diff
@@ -35,6 +35,22 @@ impl TsFn {
...
+      .unbounded_send(ThreadSafeFunctionStatus::Alive)
+      .map_err(|_| Error::GenericFailure)?;
+    Ok(())
+  }
+
+  pub fn unref(&mut self) -> Result {
+    self
+      .tsfn_sender
+      .unbounded_send(ThreadSafeFunctionStatus::Dead)
+      .map_err(|_| Error::GenericFailure)?;
```
Comment (bartlomieju) [inline cli/napi/threadsafe_functions.rs:50]: This is brittle and not idempotent - calling `unref()` multiple times will make `NapiState.active_threadsafe_functions` go negative. I was looking into fixing this and I believe the most bulletproof solution would either be implementing something similar to what @aapoalas did for FFI functions, or make `TsFn` have ID and store a hashmap of TsFn mappings in `NapiState` that keeps track of reference counts. WDYT?

**Comment by bartlomieju** [comment]
> > AFAIU ref/unref do the same thing as release/aquire functions but the later also consumes/takes ownership of the tsfn handle.
> 
> In Node `ref`/`unref` is completely unrelated to `acquire`/`release` API — they do not affect `ThreadsafeFunction` reference counter at all. [Source for `unref`](https://github.com/nodejs/node/blob/main/src/node_api.cc#L316) vs [`release`](https://github.com/nodejs/node/blob/main/src/node_api.cc#L254).

Thanks for pointing this out, we discussed this offline earlier today and decided that this integration needs to be mostly rewritten. I'll look into that.

**Review by bartlomieju — APPROVED** [review]
LGTM

**Pull request merged**